torchinductor/scheduler.py

```diff
@@ -459,87 +537,237 @@ def create_fx_from_buffers(nodes, fname, print_graph=False):
+        if isinstance(snode, FusedSchedulerNode):
+            deps = [x.node.get_reads() for x in snode.snodes]
+            deps = [x for x in itertools.chain.from_iterable(deps)]
+        else:
+            deps = node.get_reads()
+        fx_node = buf_to_fx_node[name]
 
         new_args = []
         for dep in deps:
-            if dep.name in name_to_fx_node:
-                dep_node = name_to_fx_node[dep.name]
+            if dep.name in buf_to_fx_node:
+                dep_node = buf_to_fx_node[dep.name]
             else:
                 with graph.inserting_before(first_node):
                     dep_node = graph.placeholder(dep.name)
-                    name_to_fx_node[dep.name] = dep_node
+                    buf_to_fx_node[dep.name] = dep_node
             new_args.append(dep_node)
 
         fx_node.args = tuple(new_args)
 
-    outputs = []
-    for _, v in name_to_fx_node.items():
-        if len(v.users) == 0:
-            outputs.append(v)
     graph.output(outputs[0] if len(outputs) == 1 else tuple(outputs))
+    return graph
 
-    if print_graph:
-        print(graph)
-    print("starting creating module")
-    gm = GraphModule({}, graph)
-    graph = legalize_graph(gm)
-    gm.graph.lint()
-    print("starting drawing")
-    draw_graph(gm, fname, clear_meta=False)
+
+def prefuse_nodes(snodes: List[BaseSchedulerNode]):
+    """
+    Groups together fusible scheduler nodes into a FusedSchedulerNode
+    Note: O(N^2) asymptotics
+    """
+    if len(snodes) == 0:
+        return snodes
+    from torch.fx.passes.tools_common import legalize_graph
+
+    def toposort(graph):
+        gm = fx.GraphModule({}, graph)
+        legalize_graph(gm)
+        return gm.graph
+
+    graph = create_fx_from_snodes(snodes)
+    graph = toposort(graph)
+
+    for idx, node in enumerate(graph.nodes):
+        node.order = idx
+
+    fusible_nodes = [
+        node
+        for node in graph.nodes
+        if node.op == "call_function" and node.meta["fusion_meta"].type == "compute"
+    ]
+
+    def is_fusible(src, dst):
+        # Finds whether there's a path from src to dst that isn't a direct edge
+        cur_nodes = collections.deque(
+            [user for user in src.users if user != dst]
+            + [user for user in dst.users if user != src]
+        )
+        vis = set()
+        while len(cur_nodes) > 0:
+            cur = cur_nodes.popleft()
+            if cur in vis:
+                continue
+            if cur == dst or cur == src:
+                return False
+            vis.add(cur)
+            for user in cur.users:
+                cur_nodes.append(user)
+        return True
+
+    def fuse_nodes(a: fx.Node, b: fx.Node):
+        a.args = tuple(set(a.args) | set(b.args) - set([a, b]))
+        b.replace_all_uses_with(a)
+        a.meta["fusion_meta"].snodes.extend(b.meta["fusion_meta"].snodes)
+        a.meta["fusion_meta"]._replace(type="fused")
+        a.name = a.name + "_" + b.name
```

Comment:
```suggestion
        a.name = a.name + "_" + b.name
        a.order = max(a.order, b.order)
```
For the optimization above to work.

Reply:
I’m not sure this is sufficient. The problem is that after merging two nodes a and b, where a.order < b.order, the new node cannot necessarily be located at *either* a or b.

a = f(input)
c = f(a)
d = f(input)
b = f(a, d)

a and b can be fused, but the resulting node cannot be at either a’s position nor b’s position.

I think it might still be possible to maintain some kind of structure that allows for more efficient queries, but it’s non-trivial I think.

Reply:
I agree "the resulting node cannot be at either a’s position nor b’s position."

Though for this optimization I think that is fine.  The combined node should never need to be placed *after* `max(a.order, b.order)`, so it is fine to short circuit for nodes beyond that point.

Reply:
hmm, I agree with that. Will add a note about that.

Reply:
Actually, I'm still not sure this works.

Say...

1. fuse(buf_1, buf_5) (`buf1_buf5` has order = 5)
2. is_fusible(buf_0, buf_2), where there's a path `buf_0 => buf_1 => buf_2`.
3. In `is_fusible`, we check if `buf1_buf5.order > max(buf_0.order, buf_2.order)`, which it is, since that's set as `max(buf1_buf5)`.

I think I'd prefer not to do this optimization for now. I think there has to be a better way of optimizing this query.

Reply:
Ah you are right.  We might need `.min_order`/`.max_order` to do it safely.  I am ok skipping.